loadbalancer/messages/messages.ipldsch

```diff
@@ -0,0 +1,46 @@
+type RequestKind enum {
+	| New ("n")
+  | Extend ("e")
+  | Terminate ("t")
+} representation string
+
+type ProtocolID string
+type RoutingRequest struct {
+   Kind RequestKind (rename "k")
+   Protocols [ProtocolID] (rename "ps")
+} representation map
+
+type ResponseCode enum {
+  | Ok ("o")
+  | Rejected ("r")
+} representation string
+
+type DateTime int
+
+type RoutingResponse struct {
+  Code ResponseCode (rename "c")
+  Message String (rename "m") # more info if rejected
```

Comment:
maybe `optional` too?

Reply:
there's probably a case for an inline union here with a success and failure subtype, keyed by code, something like:

```ipldsch
type RoutingResponse union {
  | RoutingResponseOk "o"
  | RoutingResponseRejected "r"
} representation inline {
  discriminantKey "c"
}

type RoutingResponseOk struct {
  Message String (rename "m")
}

type RoutingResponseRejected struct {
  Expiry DateTime
}
```

... having written all of that; I don't think this is supported in bindnode, maybe it's not even in the current Go schema parser! I guess there's a TODO here because the union would give you good validation on the existence of the two different properties.

Reply:
I believe that Union types are supported by the parser (https://github.com/ipld/go-ipld-prime/blob/7548eb883bda4712355797547a0628a0ad1c00cb/node/bindnode/infer.go#L221) but the implementation in go is pretty rough -- a wrapper struct with a set of members that are all nullable for each potential type in the union. Eek! Not sure what type safety you actually get out of that at least on the go-side, but seems like there are lots of ways to make said struct fail when actually serialized. It is enough that it makes me not want to use it.